orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java

```diff
@@ -134,7 +134,13 @@ private static VectorizedRowBatchIterator newOrcIterator(
 
     try {
       return new VectorizedRowBatchIterator(
-          file.location(), readerSchema, orcFileReader.rows(options), recordsPerBatch);
+          file.location(), readerSchema, orcFileReader.rows(options), recordsPerBatch) {
+        @Override
+        public void close() throws IOException {
```

Comment:
In the `ParquetInterable` class  we use the `CloseableGroup` `addCloseable` to make sure everything is closed

Reply:
OrcIterable also extends CloseableGroup and adds orcFileReader to closeable, but it never gets invoked. MergeIterator is operating on a CloseableIterator that might have filter and transformations and losses the closeable.

In case of `ParquetInterable` we close the reader.
````
public void close() throws IOException {
      ParquetReader.close();
    }
  }
````
but that's not. the case for OrcIterable that uses `VectorizedRowBatchIterator`
````
 @Override
  public void close() throws IOException {
    orcFileReader.rows(options).close();
  }
````

Reply:
````
CloseableIterable<>hasNext() -> inner{FilterIterator}.hasNext() -> FilterIterator.close() -> OrcIterable.OrcRowIterator.close() -> VectorizedRowBatchIterator.close()
````

Reply:
Thanks for the explanation.
So basically the `addCloseable(reader);` is not used here, neither in `ParquetIterable`, as the returned `CloseableIterator` `close` method is used clean up the resources.
And in case of the Vectorized ORC reads, we return not the `OrcRowIterator` which closes the reader, but the `VectorizedRowBatchIterator` which closes the `RecordReader rows` which is generated like this: `orcFileReader.rows(options)`.  And the `RecordReaderImpl` generated this way does not close the underlying `ReaderImpl`. 

Am I right here? Could we eventually fix this in ORC too?
If so, it would be nice to have a ORC jira saying that this could be removed once it is fixed in ORC...

Reply:
<img width="504" alt="Screenshot 2023-12-14 at 18 25 54" src="https://github.com/apache/iceberg/assets/2761528/d2f65f7d-cc22-4524-98c2-47d89aac04be">

Reply:
> Thanks for the explanation. So basically the `addCloseable(reader);` is not used here, neither in `ParquetIterable`, as the returned `CloseableIterator` `close` method is used clean up the resources. And in case of the Vectorized ORC reads, we return not the `OrcRowIterator` which closes the reader, but the `VectorizedRowBatchIterator` which closes the `RecordReader rows` which is generated like this: `orcFileReader.rows(options)`. And the `RecordReaderImpl` generated this way does not close the underlying `ReaderImpl`.
> 
> Am I right here? Could we eventually fix this in ORC too? If so, it would be nice to have a ORC jira saying that this could be removed once it is fixed in ORC...

Yes, see [HIVE-25736](https://issues.apache.org/jira/browse/HIVE-25736) :)

Reply:
Based on this, I would think that this does not affect Apache Iceberg:
https://github.com/apache/iceberg/blob/c6bbbdbc11d42713c86dd52185f577b81d946963/gradle/libs.versions.toml#L71

We use newer ORC version, which should contain ORC-498.

Maybe we still miss something.

Reply:
Ohh.. I understand it now! The ORC "fix" added the feature, that we need to close the reader explicitly...

Reply:
Any chance for a test?

Reply:
Any chance for a test?